File: txt2fpb.php
#!/usr/bin/php -q
<?php
/*

txt2fpb

This script is a complete rewrite of my old script_convert.php for Magna Carta.
No old code was used.

One day, I looked at the script and realized -- it was so sloppy I no longer
had ANY idea what the hell it was doing. Sure it worked, but it would be a cold
day in hell before anyone learned anything from it.

This script just converts the .txt script package for Magna Carta that I
released back into useable game files.

Version:   1.2a
Author:    Derrick Sobodash <derrick@sobodash.com>
Copyright: (c) 2003, 2004, 2012 Derrick Sobodash
Web site:  https://github.com/sobodash/magnacarta/
License:   BSD License <http://opensource.org/licenses/bsd-license.php>

*/

echo ("txt2fpb 1.2a (cli)\nCopyright (c) 2003, 2012 Derrick Sobodash\n");

set_time_limit(6000000);

if ($argc < 3) { DisplayOptions(); die; }
else { $path = $argv[1]; $out_path = $argv[2]; }

// Simple routine to read in a directory listing and split it to an array
$mydir=""; $outdir="";
if ($handle = opendir($path)) {
	while (false !== ($file = readdir($handle))) { 
		$mydir .= $path . "/$file\n";
		$outdir .= $out_path . "/" . substr($file, 0, strlen($file)-4) . ".fpb\n";
	}
	closedir($handle);
}
$filelist = explode("\n", $mydir); $out_list = explode("\n", $outdir);
$i=0; unset($mydir); @mkdir($out_path);

for ($z=2; $z < (count($filelist)-1); $z++) {
	unset($string_tags);

	print "Loading $filelist[$z]...\n";
	$file = trim(file_get_contents($filelist[$z]));

	print "  Parsing source files ...\n";
	$kor = fopen("./fpb/ko/" . str_replace(".txt", ".fpb", substr($filelist[$z], strlen($filelist[$z])-8, 8)), "rb");
	list($null, $kor_count) = unpack("V", fread($kor, 4));
	$lele = unpack("V*", fread($kor, 4 * 3 * $kor_count));
	print "  Harvesting string IDs ...\n";
	for($i=1; $i<count($lele); $i += 3)
		$string_tags[] = $lele[$i];

	print "  Converting text to FPB string library ...\n";
	$line_array = explode("<>", $file);
	unset($file);
	print "  Sanitizing ...\n";
	for($i=0; $i<count($line_array); $i++)
		$line_array[$i] = trim($line_array[$i]);
	unset($line_array[(count($line_array)-1)]);
	$header = pack("V*", (count($line_array)));
	$sub_off = 0;

	if(count($line_array) != count($string_tags))
	{
		echo "FATAL ERROR: The number of lines in the Korean and English files does not match! (KO=" . count($string_tags) . " EN=" . count($line_array) . ")\n";
		die;
	}

	for ($i=0; $i < count($line_array); $i++) {
		$header .= pack("V*", $string_tags[$i]);
		$header .= pack("V*", $sub_off);
		$header .= pack("V*", (strlen(str_replace("\r\n", "\$n", $line_array[$i]))+1));
		$sub_off += strlen($line_array[$i])+1;
	}
	$header .= pack("V*", $sub_off);
	unset($sub_off); $body = "";
	for ($i=0; $i < count($line_array); $i++) {
		$body .= str_replace("\r\n", "\$n", $line_array[$i]) . chr(0);
	}
	print "Writing $out_list[$z]...\n";
	$fo = fopen($out_list[$z], "w");
	fputs($fo, ($header . $body));
	fclose($fo);
}

echo ("All done!\n\n");

function DisplayOptions() {
	echo ("Builds a new set of FPB files for Magna Carta from a folder of plain text\n  usage: fpb2txt [input_path] [output_path]\n\n");
}

?>

File: txt2fpb.rb
#!/usr/bin/env ruby
#
# txt2fpb.rb
#
# This script converts the translation text files(.txt) into the (.fpb) files loaded by the game.
# Put translation files into ./script/en/
# Separate dialogues(a.k.a. "lines") in the translation files by using "<>".  Whitespace between the dialogues will get trimmed.
# Use newlines (\r\n) or "$n" to tell the game to split the dialogue onto a new line in the dialogue box.
#
# Usage:
#   ./txt2fpb.rb


puts "Starting txt2fpb.rb..."

# TODO: Translate for every language folder in ./script/
SCRIPT_DIR = "./script/en"
FPB_DIR = "./fpb/en"
KOREAN_FPB_DIR = "./fpb/ko"

#puts "SCRIPT_DIR #{SCRIPT_DIR}"
#puts "FPB_DIR #{FPB_DIR}"
#puts "KOREAN_FPB_DIR #{KOREAN_FPB_DIR}"

translation_files = Dir["#{SCRIPT_DIR}/0*.txt"] # e.g. ['script/en/0092.txt']

translation_files.each do |translation_filepath|
  puts "Loading #{translation_filepath}..."
  translation_filename = File.basename(translation_filepath)
  fpb_filename = translation_filename.sub('.txt','.fpb')
  translation_contents = File.open(translation_filepath,"r:ISO-8859-1", &:read)
  ko_file_contents = File.open("#{KOREAN_FPB_DIR}/#{fpb_filename}","r:ISO-8859-1", &:read)

#  puts "  Parsing source files..."
  first_4_bytes = ko_file_contents[0..3]
  dialogue_count = first_4_bytes.unpack('V').first
  total_header_length = 3 * 4 * dialogue_count # 3x 32-bit header bytes, each being 4x 8-bit bytes
  ko_line_headers = ko_file_contents[4..(total_header_length+4)].unpack('V*')
  string_tags = []

#  puts "  Harvesting string IDs..."
  i = 0
  while i < ko_line_headers.count
    string_tags.push(ko_line_headers[i]) # The string tags always end up being 0,1,2..etc, but we'll parse them from the korean FPB file just in case.
    i += 3
  end

#  puts "  Converting text to FPB string library..."
  line_array = translation_contents.strip.split('<>')

#  puts "  Sanitizing..."
  line_array.map!{|line|line.strip}
  header = [line_array.count].pack("V*").force_encoding("ISO-8859-1")
  sub_off = 0

  if line_array.count != string_tags.count
    raise StandardError.new("FATAL ERROR: The number of translation lines in the Korean (#{KOREAN_FPB_DIR}/#{fpb_filename}) and translation file (#{SCRIPT_DIR}/#{translation_filename}) does not match! (KO='#{string_tags.count}' EN='#{line_array.count}')")
  end

  line_array.each_with_index do |line,i|
    header += [string_tags[i]].pack("V*").force_encoding("ISO-8859-1")
    header += [sub_off].pack("V*").force_encoding("ISO-8859-1")
    header += [line.gsub("\r\n","$n").length+1].pack("V*").force_encoding("ISO-8859-1")
    sub_off += line.length+1
  end
  header += [sub_off].pack("V*").force_encoding("ISO-8859-1")
  body = "".force_encoding("ISO-8859-1")
  line_array.each do |line|
    body += (line.gsub("\r\n","$n") + 0.chr).force_encoding("ISO-8859-1")
  end
  output_file_path = "#{FPB_DIR}/#{fpb_filename}" # e.g. ./fpb/en/0134.fpb

  puts "Writing #{output_file_path}..."
  File.open(output_file_path,"w:ISO-8859-1") do |f|
    f.write(header)
    f.write(body)
  end
end


puts "Finished txt2fpb.rb\n"

